File: account-service/src/main/java/com/wks/servicemarketplace/accountservice/config/TokenValidatorFactory.java
package com.wks.servicemarketplace.accountservice.config;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.wks.servicemarketplace.accountservice.adapters.auth.TokenValidator;
import com.wks.servicemarketplace.accountservice.adapters.auth.keycloak.KeycloakJwksValidator;
import org.glassfish.hk2.api.Factory;

import javax.inject.Inject;

public class TokenValidatorFactory implements Factory<TokenValidator> {

    private final KeycloakJwksValidator tokenValidator;

    @Inject
    public TokenValidatorFactory(ApplicationParameters applicationParameters,
                                 ObjectMapper objectMapper) throws Exception {
        tokenValidator = KeycloakJwksValidator.withKeySetFromURL(
                applicationParameters.getKeycloakKeySetUrl(),
                objectMapper
        );
    }

    @Override
    public TokenValidator provide() {
        return tokenValidator;
    }

    @Override
    public void dispose(TokenValidator instance) {

    }
}

File: service-provider-service/src/main/kotlin/com/wks/servicemarketplace/serviceproviderservice/config/TokenValidatorFactory.kt
package com.wks.servicemarketplace.serviceproviderservice.config

import com.wks.servicemarketplace.serviceproviderservice.adapters.auth.StandardTokenValidator
import com.wks.servicemarketplace.serviceproviderservice.adapters.auth.TokenValidator
import org.glassfish.hk2.api.Factory
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.security.KeyFactory
import java.security.NoSuchAlgorithmException
import java.security.PublicKey
import java.security.spec.InvalidKeySpecException
import java.security.spec.X509EncodedKeySpec
import java.util.*
import java.util.stream.Collectors
import javax.inject.Inject

class TokenValidatorFactory @Inject constructor() : Factory<TokenValidator> {
    private val tokenValidator: StandardTokenValidator
    private val publicKey: PublicKey

    init {
        try {
            InputStreamReader(javaClass.classLoader.getResourceAsStream("publicKey.pem")).use { inputStreamReader ->
                BufferedReader(inputStreamReader).use { reader ->
                    val content = reader.lines().collect(Collectors.joining(""))
                            .replace("-----BEGIN PUBLIC KEY-----", "")
                            .replace("-----END PUBLIC KEY-----", "")
                    val keyFactory = KeyFactory.getInstance("RSA")
                    val keySpec = X509EncodedKeySpec(Base64.getMimeDecoder().decode(content))
                    publicKey = keyFactory.generatePublic(keySpec)
                    tokenValidator = StandardTokenValidator(publicKey)
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
            throw RuntimeException("Failed to load public key", e)
        } catch (e: NoSuchAlgorithmException) {
            e.printStackTrace()
            throw RuntimeException("Failed to load public key", e)
        } catch (e: InvalidKeySpecException) {
            e.printStackTrace()
            throw RuntimeException("Failed to load public key", e)
        }
    }

    override fun provide(): TokenValidator {
        return tokenValidator
    }

    override fun dispose(instance: TokenValidator) {}
}